File: src/features/tests/benchmark/ground-truth.ts
import type { TestKeyframe, TestLayoutIntent, TestTarget } from "../types";
import { clampTargetRect, finalizeTargetRect, TARGET_ASPECT } from "./target-geometry";

function cloneTarget(target: TestTarget): TestTarget {
  return { ...target };
}

export function layoutIntentAt(keyframe: TestKeyframe): TestLayoutIntent {
  return keyframe.layoutIntent ?? "crop";
}

/** Holds the previous keyframe intent until the next keyframe changes it. */
export function evaluateLayoutIntent(
  keyframes: TestKeyframe[],
  timestampUs: number,
): TestLayoutIntent {
  if (keyframes.length === 0) return "crop";
  if (timestampUs <= keyframes[0]!.timestampUs) return layoutIntentAt(keyframes[0]!);
  const last = keyframes.at(-1)!;
  if (timestampUs >= last.timestampUs) return layoutIntentAt(last);

  let nextIndex = 1;
  while (nextIndex < keyframes.length && keyframes[nextIndex]!.timestampUs < timestampUs) {
    nextIndex += 1;
  }
  const next = keyframes[nextIndex]!;
  if (next.timestampUs === timestampUs) return layoutIntentAt(next);
  return layoutIntentAt(keyframes[nextIndex - 1]!);
}

function interpolateCropTargets(
  previous: TestKeyframe,
  next: TestKeyframe,
  factor: number,
  sourceWidth?: number,
  sourceHeight?: number,
): TestTarget[] {
  if (layoutIntentAt(next) !== "crop") {
    return previous.targets.map(cloneTarget);
  }
  return previous.targets.map((target) => {
    const targetNext = next.targets.find((candidate) => candidate.slot === target.slot);
    if (!targetNext) return cloneTarget(target);
    if (sourceWidth && sourceHeight) {
      const centerX = target.x + target.width / 2;
      const centerY = target.y + target.height / 2;
      const nextCenterX = targetNext.x + targetNext.width / 2;
      const nextCenterY = targetNext.y + targetNext.height / 2;
      const height = target.height + (targetNext.height - target.height) * factor;
      const width = (height * sourceHeight * TARGET_ASPECT) / sourceWidth;
      return {
        ...target,
        ...finalizeTargetRect({
          ...target,
          x: centerX + (nextCenterX - centerX) * factor - width / 2,
          y: centerY + (nextCenterY - centerY) * factor - height / 2,
          width,
          height,
        }, sourceWidth, sourceHeight),
      };
    }
    return {
      ...target,
      ...clampTargetRect({
        ...target,
        x: target.x + (targetNext.x - target.x) * factor,
        y: target.y + (targetNext.y - target.y) * factor,
        width: target.width + (targetNext.width - target.width) * factor,
        height: target.height + (targetNext.height - target.height) * factor,
      }),
    };
  });
}

function interpolateContainTargets(
  previous: TestKeyframe,
  next: TestKeyframe,
  factor: number,
): TestTarget[] {
  const target = previous.targets[0];
  if (!target) return [];
  const targetNext = layoutIntentAt(next) === "contain" ? next.targets[0] : undefined;
  if (!targetNext) return [cloneTarget(target)];
  return [{
    ...target,
    ...clampTargetRect({
      x: target.x + (targetNext.x - target.x) * factor,
      y: target.y + (targetNext.y - target.y) * factor,
      width: target.width + (targetNext.width - target.width) * factor,
      height: target.height + (targetNext.height - target.height) * factor,
    }),
    slot: 0,
  }];
}

/** Evaluates the manual timeline. Endpoint values are held; target-count changes occur on keyframes. */
export function evaluateGroundTruth(
  keyframes: TestKeyframe[],
  timestampUs: number,
  sourceWidth?: number,
  sourceHeight?: number,
): TestTarget[] {
  if (keyframes.length === 0) return [];
  if (timestampUs <= keyframes[0]!.timestampUs) {
    return keyframes[0]!.targets.map(cloneTarget);
  }
  const last = keyframes.at(-1)!;
  if (timestampUs >= last.timestampUs) return last.targets.map(cloneTarget);

  let nextIndex = 1;
  while (nextIndex < keyframes.length && keyframes[nextIndex]!.timestampUs < timestampUs) {
    nextIndex += 1;
  }
  const next = keyframes[nextIndex]!;
  if (next.timestampUs === timestampUs) return next.targets.map(cloneTarget);
  const previous = keyframes[nextIndex - 1]!;
  const factor = (timestampUs - previous.timestampUs) /
    Math.max(1, next.timestampUs - previous.timestampUs);

  if (evaluateLayoutIntent(keyframes, timestampUs) === "contain") {
    return interpolateContainTargets(previous, next, factor);
  }
  return interpolateCropTargets(previous, next, factor, sourceWidth, sourceHeight);
}

export function normalizeKeyframes(
  keyframes: TestKeyframe[],
  sourceWidth?: number,
  sourceHeight?: number,
): TestKeyframe[] {
  return [...keyframes]
    .map((frame) => {
      const layoutIntent = frame.layoutIntent ?? "crop";
      return {
        ...frame,
        layoutIntent,
        timestampUs: Math.max(0, Math.round(frame.timestampUs)),
        targets: [...frame.targets]
          .sort((a, b) => a.slot - b.slot)
          .map((target, index) => {
            const slot = index as 0 | 1;
            const geometry = layoutIntent === "contain"
              ? clampTargetRect({ ...target, slot })
              : sourceWidth && sourceHeight
                ? finalizeTargetRect({ ...target, slot }, sourceWidth, sourceHeight)
                : clampTargetRect({ ...target, slot });
            return { ...target, ...geometry, slot };
          }),
      };
    })
    .sort((a, b) => a.timestampUs - b.timestampUs);
}

File: src-tauri/src/commands/benchmark_miss_export/ground_truth.rs
use crate::repository::test_repository::{TestKeyframeDto, TestTargetDto};

fn clone_target(target: &TestTargetDto) -> TestTargetDto {
    TestTargetDto {
        id: target.id.clone(),
        slot: target.slot,
        x: target.x,
        y: target.y,
        width: target.width,
        height: target.height,
    }
}

fn layout_intent_at(frame: &TestKeyframeDto) -> &str {
    if frame.layout_intent.is_empty() || frame.layout_intent == "crop" {
        "crop"
    } else {
        frame.layout_intent.as_str()
    }
}

enum KeyframeBracket<'a> {
    BeforeFirst(&'a TestKeyframeDto),
    AfterLast(&'a TestKeyframeDto),
    Exact(&'a TestKeyframeDto),
    Between {
        previous: &'a TestKeyframeDto,
        next: &'a TestKeyframeDto,
        factor: f64,
    },
}

fn keyframe_bracket(keyframes: &[TestKeyframeDto], timestamp_us: i64) -> Option<KeyframeBracket<'_>> {
    if keyframes.is_empty() {
        return None;
    }
    if timestamp_us <= keyframes[0].timestamp_us {
        return Some(KeyframeBracket::BeforeFirst(&keyframes[0]));
    }
    let last = keyframes.last().expect("keyframes checked");
    if timestamp_us >= last.timestamp_us {
        return Some(KeyframeBracket::AfterLast(last));
    }
    let mut next_index = 1usize;
    while next_index < keyframes.len() && keyframes[next_index].timestamp_us < timestamp_us {
        next_index += 1;
    }
    let next = &keyframes[next_index];
    if next.timestamp_us == timestamp_us {
        return Some(KeyframeBracket::Exact(next));
    }
    let previous = &keyframes[next_index - 1];
    let factor = (timestamp_us - previous.timestamp_us) as f64
        / (next.timestamp_us - previous.timestamp_us).max(1) as f64;
    Some(KeyframeBracket::Between {
        previous,
        next,
        factor,
    })
}

fn evaluate_layout_intent(keyframes: &[TestKeyframeDto], timestamp_us: i64) -> &str {
    match keyframe_bracket(keyframes, timestamp_us) {
        None => "crop",
        Some(KeyframeBracket::BeforeFirst(frame))
        | Some(KeyframeBracket::AfterLast(frame))
        | Some(KeyframeBracket::Exact(frame)) => layout_intent_at(frame),
        Some(KeyframeBracket::Between { previous, .. }) => layout_intent_at(previous),
    }
}

fn clamp_target_rect(target: &TestTargetDto) -> TestTargetDto {
    let width = target.width.clamp(0.001, 1.0);
    let height = target.height.clamp(0.001, 1.0);
    TestTargetDto {
        id: target.id.clone(),
        slot: target.slot,
        x: target.x.clamp(0.0, 1.0 - width),
        y: target.y.clamp(0.0, 1.0 - height),
        width,
        height,
    }
}

fn interpolate_contain_targets(
    previous: &TestKeyframeDto,
    next: &TestKeyframeDto,
    factor: f64,
) -> Vec<TestTargetDto> {
    let Some(target) = previous.targets.first() else {
        return Vec::new();
    };
    let target_next = if layout_intent_at(next) == "contain" {
        next.targets.first()
    } else {
        None
    };
    let Some(target_next) = target_next else {
        return vec![clone_target(target)];
    };
    vec![clamp_target_rect(&TestTargetDto {
        id: target.id.clone(),
        slot: 0,
        x: target.x + (target_next.x - target.x) * factor,
        y: target.y + (target_next.y - target.y) * factor,
        width: target.width + (target_next.width - target.width) * factor,
        height: target.height + (target_next.height - target.height) * factor,
    })]
}

fn interpolate_crop_targets(
    previous: &TestKeyframeDto,
    next: &TestKeyframeDto,
    factor: f64,
) -> Vec<TestTargetDto> {
    if layout_intent_at(next) != "crop" {
        return previous.targets.iter().map(clone_target).collect();
    }
    previous
        .targets
        .iter()
        .map(|target| {
            let target_next = next.targets.iter().find(|candidate| candidate.slot == target.slot);
            match target_next {
                Some(next_target) => TestTargetDto {
                    id: target.id.clone(),
                    slot: target.slot,
                    x: target.x + (next_target.x - target.x) * factor,
                    y: target.y + (next_target.y - target.y) * factor,
                    width: target.width + (next_target.width - target.width) * factor,
                    height: target.height + (next_target.height - target.height) * factor,
                },
                None => clone_target(target),
            }
        })
        .collect()
}

pub(crate) fn evaluate_ground_truth(
    keyframes: &[TestKeyframeDto],
    timestamp_us: i64,
) -> Vec<TestTargetDto> {
    match keyframe_bracket(keyframes, timestamp_us) {
        None => Vec::new(),
        Some(KeyframeBracket::BeforeFirst(frame) | KeyframeBracket::Exact(frame)) => {
            frame.targets.iter().map(clone_target).collect()
        }
        Some(KeyframeBracket::AfterLast(frame)) => frame.targets.iter().map(clone_target).collect(),
        Some(KeyframeBracket::Between {
            previous,
            next,
            factor,
        }) => {
            if evaluate_layout_intent(keyframes, timestamp_us) == "contain" {
                interpolate_contain_targets(previous, next, factor)
            } else {
                interpolate_crop_targets(previous, next, factor)
            }
        }
    }
}
